src/main/scala/mesosphere/marathon/storage/migration/MigrationTo18.scala

```diff
@@ -0,0 +1,175 @@
+      _.map { case (k, v) => Task.Id(k) -> v }
+    }
+  }
+
+  /**
+    * Read format for old instance without goal.
+    */
+  val instanceJsonReads17: Reads[Instance] = {
+    (
+      (__ \ "instanceId").read[Instance.Id] ~
+      (__ \ "agentInfo").read[AgentInfo] ~
+      (__ \ "tasksMap").read[Map[Task.Id, Task]](taskMapReads17) ~
+      (__ \ "runSpecVersion").read[Timestamp] ~
+      (__ \ "state").read[InstanceState](instanceStateReads17) ~
+      (__ \ "unreachableStrategy").readNullable[raml.UnreachableStrategy] ~
+      (__ \ "reservation").readNullable[Reservation]
+    ) { (instanceId, agentInfo, tasksMap, runSpecVersion, state, ramlUnreachableStrategy, reservation) =>
+        val unreachableStrategy = ramlUnreachableStrategy.
+          map(Raml.fromRaml(_)).getOrElse(UnreachableStrategy.default())
+        new Instance(instanceId, Some(agentInfo), state, tasksMap, runSpecVersion, unreachableStrategy, reservation)
+      }
+  }
+
+  implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+    new IdResolver[Instance.Id, JsValue, String, ZkId] {
+      override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+        ZkId(category, id.idString, version)
+
+      override val category: String = "instance"
+
+      override def fromStorageId(key: ZkId): Id = Instance.Id.fromIdString(key.id)
+
+      override val hasVersions: Boolean = false
+
+      override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+    }
+
+  implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+    Unmarshaller.strict {
+      case ZkSerialized(byteString) =>
+        Json.parse(byteString.utf8String)
+    }
+
+  /**
+    * This function traverses all instances in ZK and sets the instance goal field.
+    */
+  def migrateInstanceConditions(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit mat: Materializer): Future[Done] = {
+
+    logger.info("Starting instance condition migration")
+
+    maybeStore(persistenceStore).map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get[Instance.Id, JsValue](instanceId)
+        }
+        .via(migrationFlow)
+        .mapAsync(1) { updatedInstance =>
+          instanceRepository.store(updatedInstance)
+        }
+        .runWith(Sink.ignore)
+    } getOrElse {
+      Future.successful(Done)
+    }
+  }
+
+  /**
+    * Extract instance from old format with replaced condition.
+    * @param jsValue The instance as JSON.
+    * @return The parsed instance.
+    */
+  def extractInstanceFromJson(jsValue: JsValue): Instance = jsValue.as[Instance](instanceJsonReads17)
+
+  // This flow parses all provided instances and updates their goals. It does not save the updated instances.
+  val migrationFlow = Flow[Option[JsValue]]
+    .mapConcat {
+      case Some(jsValue) =>
+        List(extractInstanceFromJson(jsValue))
```

Comment:
Hm, I'm not sure this is a common pattern in Akka streams but I would avoid the allocation of lists and simply flatten the stream before
```suggestion
  .mapConcat(id)
  .map(extractInstanceFromJson)
```

Reply:
the problem is that scala `Option` is not an `Iterable`, and we still need to transform it into List.